### PR TITLE
Industry Window Implementation

### PR DESCRIPTION
**Pull request by Svelbeard**

### CODE REVIEW

File: openloco.sln [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
-# Visual Studio 15
-VisualStudioVersion = 15.0.27130.2010
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.30011.22
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "openloco", "src\openloco\openloco.vcxproj", "{42A6B551-4EC5-4B66-A130-628622CD98C4}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{D35A17D9-494B-43DB-8008-B47DC58227DD}"
+	ProjectSection(SolutionItems) = preProject
+		src\openloco\windows\IndustryWindow.cpp = src\openloco\windows\IndustryWindow.cpp
```
Comment (AaronVanGeffen) [inline openloco.sln:10]: Please revert the changes to this file.

File: src/openloco/openloco.vcxproj [hunk 2]
```diff
@@ -15,7 +15,110 @@
     </ProjectConfiguration>
   </ItemGroup>
   <ItemGroup>
-    <ClCompile Include="**\*.cpp" Exclude="version.cpp" />
+    <ClCompile Include="audio\audio.cpp" />
```
Comment (AaronVanGeffen) [inline src/openloco/openloco.vcxproj:18]: Please revert the (accidental?) changes to this file. We use wildcards on purpose.

File: src/openloco/ui/WindowManager.h [hunk 3]
```diff
@@ -97,6 +97,11 @@ namespace openloco::ui::windows::construction
     void on_mouse_up(window& w, uint16_t widgetIndex);
 }
 
+namespace openloco::ui::windows::industry
+{
+    window* open(uint16_t id);
```
Comment (AaronVanGeffen) [inline src/openloco/ui/WindowManager.h:102]: Please add an `industry_id_t` to types.hpp.

File: src/openloco/windows/IndustryWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+            {
+                window->colours[0] = skin->colour_0B;
+                window->colours[1] = skin->colour_0C;
+            }
+            // 0x00456DBC end
+
+            window->saved_view.clear();
+        }
+
+        // TODO(avgeffen): only needs to be called once.
```
Comment (AaronVanGeffen) [inline src/openloco/windows/IndustryWindow.cpp:267]: This is your line, not mine. You can leave out the name, though, as we don't really follow the convention in this project. Just write:
```suggestion
        // TODO: only needs to be called once.
```

**Comment by AaronVanGeffen** [comment]
Thanks for the PR. Nice to see you've gotten a taste for it!

It looks like you accidentally committed some changes to the Visual Studio project files, though. Could you revert these?

File: openloco.sln [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
+	EndProjectSection
 EndProject
```
Comment (Svelbeard) [inline openloco.sln]: I'm not sure I added a file to the project correctly, could someone confirm?

File: src/openloco/openloco.vcxproj (hunk 2 above)
Comment (Svelbeard) [reply src/openloco/openloco.vcxproj]: Sorry I hadn't noticed that I'd commited the changes to this

File: src/openloco/windows/IndustryWindow.cpp (hunk 4 above)
Comment (Svelbeard) [reply src/openloco/windows/IndustryWindow.cpp:267]: OK, I hadn't realised it was your name and was wondering what it stood for! I think I did the same in Station Window PR.

**Comment by Svelbeard** [comment]
Yep all done, sorry about that.

File: src/openloco/industry.cpp [hunk 6]
```diff
@@ -111,13 +111,13 @@ namespace openloco
         }
     }
 
-    void industry::sub_454A43(coord_t x, coord_t y, uint8_t bl, uint8_t bh, uint8_t dl)
+    void industry::sub_454A43(coord_t ax, coord_t cx, uint8_t bl, uint8_t bh, uint8_t dl)
```
Comment (duncanspumpkin) [inline src/openloco/industry.cpp:114]: ax and cx are still an x and y but they represent a random tile near the industry item. I would like to see this replaced with a `map::map_pos` as it is a reference to a tile.

File: src/openloco/openloco.vcxproj [hunk 7]
```diff
@@ -67,4 +67,4 @@
     <_NuGetTargetFallbackMoniker>$(_NuGetTargetFallbackMoniker);native,Version=v0.0</_NuGetTargetFallbackMoniker>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
-</Project>
```
Comment (duncanspumpkin) [inline src/openloco/openloco.vcxproj:70]: revert

File: src/openloco/windows/IndustryWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+                self->widgets[i].right = xPos + tabWidth;
+                xPos = self->widgets[i].right + 1;
+            }
+        }
+        static void initEvents()
+        {
+            industry::initEvents();
+        }
+    }
+}
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:414]: add a return

File: src/openloco/industry.cpp (hunk 6 above)
Comment (Svelbeard) [reply src/openloco/industry.cpp:114]: like this?
```
void industry::sub_454A43(map_pos pos, uint8_t bl, uint8_t bh, uint8_t dl)
{
    registers regs;
    regs.bl = bl;
    regs.bh = bh;
    regs.ax = pos.x;
    regs.cx = pos.y;
    regs.dl = dl;
    regs.dh = id();
    call(0x00454A43, regs);
}
```

Comment (duncanspumpkin) [reply src/openloco/industry.cpp:114]: Yup that looks better

File: src/openloco/game_commands.h [hunk 9]
```diff
@@ -157,6 +157,28 @@ namespace openloco::game_commands
...
+        regs.edi = edi; // part of name buffer
+        do_command(79, regs);
+    }
+
+    inline bool do_48(uint8_t bl, uint8_t industryId)
+    {
+        registers regs;
+        regs.bl = bl; // [ 1 = remove industry]
+        regs.edi = industryId;
+        do_command(50, regs);
```
Comment (Svelbeard) [inline src/openloco/game_commands.h:178]: should be `do_command(48, regs);`

File: src/openloco/window.h [hunk 10]
```diff
@@ -27,7 +27,7 @@ namespace openloco::ui
     {
         widget_type type; // 0x00
         uint8_t colour;   // 0x01
-        int16_t left;     // 0x02
+        uint16_t left;    // 0x02
```
Comment (Svelbeard) [inline src/openloco/window.h:30]: I changed this so that the widget->left could be set to 0x8000. Which is then checked in drawTabs to see if the widget is disabled. But this causes the CI checks to fail due to `if (widget.left != -2)` in WIndowManager.cpp being always true. I will revert it and check for disabled widgets instead.

File: src/openloco/widget.cpp [hunk 11]
```diff
@@ -786,6 +786,7 @@ namespace openloco::ui::widget
 
         if (w->is_disabled(index))
         {
+            widget->left = 0x8000;
```
Comment (Svelbeard) [inline src/openloco/widget.cpp:789]: I changed this so that the widget->left could be set to 0x8000. Which is then checked in drawTabs to see if the widget is disabled. But this causes the CI checks to fail due to if (widget.left != -2) in WIndowManager.cpp being always true. I will revert it and check for disabled widgets instead.

**Comment by Svelbeard** [comment]
Would someone mind reviewing this? I could use some of these changes in the IndustryList PR #457.
Thanks.

**Comment by duncanspumpkin** [comment]
I'm halfway through reviewing it but github doesn't show that until I'm complete.

File: src/openloco/windows/IndustryWindow.cpp [hunk 12]
```diff
@@ -0,0 +1,1064 @@
...
+            auto industryObj = objectmgr::get<industry_object>(industry->object_id);
+            auto cargoObj = objectmgr::get<cargo_object>((uint8_t)industryObj->produced_cargo[0]);
+
+            auto args = FormatArguments();
+            args.push(cargoObj->units_and_cargo_name);
+
+            int16_t x = self->x + 2;
+            int16_t y = self->y - 24 + 68;
+
+            gfx::draw_string_494B3F(*dpi, x, y, colour::black, string_ids::production_graph_label, &args);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:398]: Put all of this  (L387 - L398) inside of curly braces so that you can reuse args and x, y variable names

File: src/openloco/windows/IndustryWindow.cpp [hunk 13]
```diff
@@ -0,0 +1,1064 @@
...
+                args2.push(yTick);
+
+                gfx::draw_rect(dpi, self->x + 41, yPos, self->x + 280, 1, colour::get_shade(self->colours[1], 4));
+
+                gfx::draw_string_494C78(*dpi, self->x + 39, yPos - 6, colour::black, string_ids::population_graph_people, &args2);
+
+                yTick += 1000;
+            }
+
+            int8_t month = static_cast<int8_t>(current_month());
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:414]: I know this just copies the town window but I'd prefer we kept it in the same type as current_month that way we can use the enum correctly. It will mean some static casts required for incrmenting/decrementing though.

File: src/openloco/windows/IndustryWindow.cpp [hunk 14]
```diff
@@ -0,0 +1,1064 @@
...
+                {
+                    if (yPos1 < self->y + self->height - 7)
+                    {
+                        if (yPos2 < self->y + self->height - 7)
+                        {
+                            gfx::draw_line(dpi, xPos, yPos1, xPos + 1, yPos2, colour::get_shade(self->colours[1], 7));
+                        }
+                    }
+                }
+                month--;
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:451]: Change this to check if january before the decrement rather than after.

File: src/openloco/windows/IndustryWindow.cpp [hunk 15]
```diff
@@ -0,0 +1,1064 @@
...
+                if (month < 0)
+                {
+                    month = 11;
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:454]: december

File: src/openloco/windows/IndustryWindow.cpp [hunk 16]
```diff
@@ -0,0 +1,1064 @@
...
+        // 0x0045630F
+        static void draw(window* self, gfx::drawpixelinfo_t* dpi)
+        {
+            self->draw(dpi);
+            common::drawTabs(self, dpi);
+
+            // Draw Units of Cargo sub title
+            auto industry = industrymgr::get(self->number);
+            auto industryObj = objectmgr::get<industry_object>(industry->object_id);
+            auto cargoObj = objectmgr::get<cargo_object>((uint8_t)industryObj->produced_cargo[1]);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:535]: See above comments. Also is there a way we can make a common function for these two tabs.

File: src/openloco/windows/IndustryWindow.cpp [hunk 17]
```diff
@@ -0,0 +1,1064 @@
...
+                    if (yearSkip >= 3)
+                        yearSkip = 0;
+                }
+            }
+        }
+
+        // 0x004565B5
+        static void on_mouse_up(window* self, widget_index widgetIndex)
+        {
+            switch (widgetIndex)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:612]: Identical to production_1 try make this the exact same common function

File: src/openloco/windows/IndustryWindow.cpp [hunk 18]
```diff
@@ -0,0 +1,1064 @@
...
+
+        // 0x00456705
+        static void draw(window* self, gfx::drawpixelinfo_t* dpi)
+        {
+            self->draw(dpi);
+            common::drawTabs(self, dpi);
+
+            auto industry = industrymgr::get(self->number);
+            auto industryObj = objectmgr::get<industry_object>(industry->object_id);
+            auto totalReceivedCargo = industryObj->received_cargo[0] & industryObj->received_cargo[1] & industryObj->received_cargo[2];
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:677]: This is really just a cheap way of checking that at least one of the received_cargo s is not null. I think making the name `canReceiveCargo` and making it a bool would be better. Same principle goes for the produced cargo.

File: src/openloco/windows/IndustryWindow.cpp [hunk 19]
```diff
@@ -0,0 +1,1064 @@
...
+            int16_t xPos = self->x + 3;
+            int16_t yPos = self->y + 45;
+            gfx::point_t origin = { xPos, yPos };
+            // Draw Last Months received cargo stats
+            if (totalReceivedCargo != 0xFF)
+            {
+                origin.x += 4;
+                origin.y += 10;
+                gfx::draw_string_494B3F(*dpi, xPos, yPos, colour::black, string_ids::received_cargo);
+                for (uint8_t i = 0; i < std::size(industryObj->received_cargo); i++)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:687]: This can become a ranged for looping over all of the received cargo types

File: src/openloco/windows/IndustryWindow.cpp [hunk 20]
```diff
@@ -0,0 +1,1064 @@
...
+                {
+                    uint16_t receivedCargo = industryObj->received_cargo[i];
+                    if (receivedCargo != 0xFF)
+                    {
+                        auto cargoObj = objectmgr::get<cargo_object>(receivedCargo);
+                        auto args = FormatArguments();
+
+                        if (receivedCargo == 1)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:695]: This is why it is important to name you're variables better. This is the `receivedCargoType` not the `receviedCargoQuantity` this line does not make sense. You will need to use the variable from L703.

File: src/openloco/windows/IndustryWindow.cpp [hunk 21]
```diff
@@ -0,0 +1,1064 @@
...
+                        }
+                        args.push(industry->received_cargo[i]);
+
+                        gfx::draw_string_left_wrapped(dpi, &origin, 290, colour::black, string_ids::white_stringid2, &args);
+                    }
+                }
+                origin.y += 4;
+            }
+
+            auto totalProducedCargo = industryObj->produced_cargo[0] & industryObj->produced_cargo[1];
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:711]: See above

File: src/openloco/windows/IndustryWindow.cpp [hunk 22]
```diff
@@ -0,0 +1,1064 @@
...
+            // Draw Last Months produced cargo stats
+            if (totalProducedCargo != 0xFF)
+            {
+                gfx::draw_string_494B3F(*dpi, origin.x - 4, origin.y, colour::black, string_ids::produced_cargo);
+                origin.y += 10;
+
+                for (uint8_t i = 0; i < std::size(industryObj->produced_cargo); i++)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:718]: See above

File: src/openloco/windows/IndustryWindow.cpp [hunk 23]
```diff
@@ -0,0 +1,1064 @@
...
+                {
+                    uint16_t producedCargo = industryObj->produced_cargo[i];
+                    if (producedCargo != 0xFF)
+                    {
+                        auto cargoObj = objectmgr::get<cargo_object>(producedCargo);
+                        auto args = FormatArguments();
+
+                        if (producedCargo == 1)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:726]: Same mistake as above

File: src/openloco/windows/IndustryWindow.cpp [hunk 24]
```diff
@@ -0,0 +1,1064 @@
...
+
+                        if (receivedCargo == 1)
+                        {
+                            args.push(cargoObj->unit_name_singular);
+                        }
+                        else
+                        {
+                            args.push(cargoObj->unit_name_plural);
+                        }
+                        args.push(industry->received_cargo[i]);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:703]: You will need to push a full 32bits here not 16 bits. You've done it correct for the produced cargo.

File: src/openloco/windows/IndustryWindow.cpp [hunk 25]
```diff
@@ -0,0 +1,1064 @@
...
+                        if (producedCargo == 1)
+                        {
+                            args.push(cargoObj->unit_name_singular);
+                        }
+                        else
+                        {
+                            args.push(cargoObj->unit_name_plural);
+                        }
+                        args.push<uint32_t>(industry->produced_cargo[i]);
+                        args.push(industry->var_1A7[i]);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:735]: var_1A7 is the percentage produced cargo transported.

File: src/openloco/windows/IndustryWindow.cpp [hunk 26]
```diff
@@ -0,0 +1,1064 @@
...
+
+                        gfx::draw_string_left_wrapped(dpi, &origin, 290, colour::black, string_ids::transported_cargo, &args);
+                    }
+                }
+            }
+        }
+
+        // 0x0045695E
+        static void on_mouse_up(window* self, widget_index widgetIndex)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:744]: This looks identical to the production tab as well.

File: src/openloco/windows/IndustryWindow.cpp [hunk 27]
```diff
@@ -0,0 +1,1064 @@
...
+            {
+                self->set_size(windowSize, windowSize);
+            }
+        }
+
+        // 0x0045698E
+        static void update(window* self)
+        {
+            self->frame_no++;
+            if (self->frame_no > 28)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:777]: `>=`

File: src/openloco/windows/IndustryWindow.cpp [hunk 28]
```diff
@@ -0,0 +1,1064 @@
...
+                self->frame_no = 0;
+
+            self->call_prepare_draw();
+            WindowManager::invalidateWidget(WindowType::industry, self->number, self->current_tab + 4);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:781]: 4 is the widx:: for the first tab

File: src/openloco/windows/IndustryWindow.cpp [hunk 29]
```diff
@@ -0,0 +1,1064 @@
...
+        // 0x004569C2
+        static void on_resize(window* self)
+        {
+            {
+                self->set_size(windowSize, windowSize);
+            }
+        }
+
+        // 0x0045698E
+        static void update(window* self)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:774]: This can actually just be removed and the common function used instead. The frame_number is checked correctly when used and the invalidateWidget is just a more efficient invalidate. I suggest we remove this function and use the common version. Would be worth checking though to make sure that is true.

File: src/openloco/windows/IndustryWindow.cpp [hunk 30]
```diff
@@ -0,0 +1,1064 @@
...
+            common::drawTabs(self, dpi);
+            self->drawViewports(dpi);
+            widget::drawViewportCentreButton(dpi, self, widx::centre_on_viewport);
+
+            //auto industry = industrymgr::get(self->number);
+
+            // 0x0045935F start
+            registers regs;
+            regs.dx = self->number;
+            call(0x0045935F, regs);
```
[the remark this answers is not in the document]
Comment (Svelbeard) [reply src/openloco/windows/IndustryWindow.cpp:143]: I've already reverse engineered this function in the IndustryList. Shall I move it to this one instead? And should I put it in industry.cpp?

File: src/openloco/windows/IndustryWindow.cpp (hunk 14 above)
Comment (Svelbeard) [reply src/openloco/windows/IndustryWindow.cpp:451]: It does already in L424.

Comment (Svelbeard) [reply src/openloco/windows/IndustryWindow.cpp:451]: Nevermind I understand what you meant now.

File: src/openloco/windows/IndustryWindow.cpp (hunk 29 above)
Comment (Svelbeard) [reply src/openloco/windows/IndustryWindow.cpp:774]: I see no difference when using the common function so will remove it.

File: src/openloco/windows/IndustryWindow.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openloco/windows/IndustryWindow.cpp:143]: Okay just leave it for the industry list pr

File: src/openloco/windows/IndustryWindow.cpp [hunk 31]
```diff
@@ -809,6 +562,131 @@ namespace openloco::ui::windows::industry
...
+                gfx::draw_string_494C78(*dpi, self->x + 39, yPos - 6, colour::black, string_ids::population_graph_people, &args);
+
+                yTick += 1000;
+            }
+
+            month_id month = current_month();
+            int16_t year = current_year();
+            int8_t yearSkip = 0;
+
+            for (uint8_t i = industry->history_size[0] - 1; i > 0; i--)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:618]: You will need to choose between history_size[0] and history_size[1] based on the current tab.

File: src/openloco/windows/IndustryWindow.cpp [hunk 32]
```diff
@@ -809,6 +562,131 @@ namespace openloco::ui::windows::industry
...
+                        auto args = FormatArguments();
+                        args.push(year);
+
+                        gfx::draw_string_centred(*dpi, xPos, yPos, colour::black, string_ids::population_graph_year, &args);
+                    }
+
+                    gfx::draw_rect(dpi, xPos, yPos + 11, 1, self->y + self->height - 7, colour::get_shade(self->colours[1], 4));
+                }
+                // Draw population graph
+                uint16_t yPos1 = -industry->history_1[i] + (self->y + self->height - 7);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:637]: choose between history_1 and history_2 based on current tab

File: src/openloco/windows/IndustryWindow.cpp [hunk 33]
```diff
@@ -809,6 +562,131 @@ namespace openloco::ui::windows::industry
...
+                uint16_t yPos2 = -industry->history_1[i + 1] + (self->y + self->height - 7);
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:638]: as above

File: src/openloco/windows/IndustryWindow.cpp [hunk 34]
```diff
@@ -809,6 +562,131 @@ namespace openloco::ui::windows::industry
...
+
+                // Do not draw current segment yet; it may be zeroed.
+                if (i < industry->history_size[0] - 1)
```
Comment (duncanspumpkin) [inline src/openloco/windows/IndustryWindow.cpp:641]: as above

File: src/openloco/industry.cpp [hunk 35]
```diff
@@ -113,7 +113,7 @@ namespace openloco
                         {
                             if (prng.rand_bool())
                             {
-                                map::map_pos randTile{ x + coord_t(prng.rand_next(-15, 16) * 32), y + coord_t(prng.rand_next(-15, 16) * 32) };
+                                map::map_pos randTile{ x + static_cast<coord_t>(prng.rand_next(-15, 16) * 32), y + static_cast<coord_t>(prng.rand_next(-15, 16) * 32) };
```
Comment (duncanspumpkin) [inline src/openloco/industry.cpp:116]: the `+` is causing this to be promoted to an int. just do a static cast of the whole thing ` static_cast<coord_t>(x + prng.rand_next(-15, 16) * 32)`

Comment (Svelbeard) [reply src/openloco/industry.cpp:116]: Thanks, I was getting frustrated with this!

**Comment by duncanspumpkin** [comment]
I noticed a few little bugs when testing and it was going to be too annoying to comment how exactly to fix it. So I've created a commit to fix them see https://github.com/duncanspumpkin/OpenLoco/commit/0b17fa3d50134e4c64ce7a8ec7cd85ca74ecb6e9 
It seems you haven't enabled the ability for me to push my changes direct to your branch.

**Comment by Svelbeard** [comment]
I've just added you and @AaronVanGeffen  as collaborators on my forked repository

**Comment by AaronVanGeffen** [comment]
You can set this permission on a PR basis with the "Allow modifications from collaborators" checkbox. It shouldn't be necessary to add us to your fork.

**Comment by Svelbeard** [comment]
I’ve cancelled the requests and checked the option for “allow edits from maintainers”

**Comment by AaronVanGeffen** [comment]
It doesn't appear to have been saved correctly, unfortunately. My `push` gets rejected when trying to push @duncanspumpkin's commit to your branch.

**Comment by duncanspumpkin** [comment]
I've noticed issues happening with edits from contributors this week so might be a github problem.
